SimplifyCFG: Enable switch replacements in more cases

OutOfCache opened the pull request:
In some cases, we can replace a switch with simpler instructions or a lookup table.
For instance, if every case results in the same value, we can simply replace the switch
with that single value.

However, lookup tables are not always supported. 
Targets, function attributes and compiler options can deactivate lookup table creation.
Currently, even simpler switch replacements like the single value optimization do not
get applied, because we only enable these transformations if lookup tables are enabled.

This PR enables the other kinds of replacements, even if lookup tables are not supported.
First, it checks if the potential replacements are lookup tables.
If they are, then check if lookup tables are supported and whether to continue.
If they are not, then we can apply the other transformations.

Originally, lookup table creation was delayed until late stages of the compilation pipeline, because
it can result in difficult-to-analyze code and prevent other optimizations.
As a side effect of this change, we can also enable the simpler optimizations much earlier in the
compilation process.

OutOfCache committed:
SimplifyCFG: Add `isLookupTable` method

Check if the SwitchReplacement is a lookup table or not.

OutOfCache committed:
SimplifyCFG: Delay check for switch-to-lookup option

The option prevents any switch replacements from happening earlier in
the compilation pipeline, even though it should only apply to lookup
tables.

Delay the check until we are sure that we would create a lookup table.
If not, continue with the replacement.

OutOfCache committed:
SimplifyCFG: Delay check for LUT support in target

Only check the target support for LUTs if we would create one.
Otherwise, proceed with the optimization.

OutOfCache committed:
SimplifyCFG: Delay check for "no-jump-tables" attribute

Allow the replacement of switches with anything other than a lookup
table, even if the "no-jump-table" attribute is set.

OutOfCache committed:
SimplifyCFG: Add tests for switch-replacements

Currently, some switch optimizations are only possible, if lookup table
support is enabled. However, some of these optimizations are no lookup
tables at all, but a series of simple instructions like mul, add
or shift instructions.

These tests will show how a future change will enable these other kinds
of optimization, even if lookup tables are not supported.

dtcxzyw commented on llvm/test/Transforms/SimplifyCFG/switch-transformations-no-lut.ll:
```diff
@@ -0,0 +1,445 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -passes='simplifycfg' < %s | FileCheck %s --check-prefix=OPTNOLUT
+; RUN: opt -mtriple=amdgcn--amdpal -S -passes='simplifycfg<switch-to-lookup>' < %s | FileCheck %s --check-prefix=TTINOLUT
```
Should be wrapped with `%if amdgpu-registered-target %{ ... %}`.

OutOfCache committed:
fixup! SimplifyCFG: Add tests for switch-replacements

OutOfCache committed:
SimplifyCFG: Delay creation of bitmaps

Early bitmap creation can, similar to lookup tables,
cause missed optimizations because they are more difficult to analyze.

Delay the creation until later in the pipeline, when the option
`switch-to-lookup` is set in in SimplifyCFG.

OutOfCache commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -7086,6 +7087,26 @@ static bool simplifySwitchLookup(SwitchInst *SI, IRBuilder<> &Builder,
+  //        pipeline, because it would otherwise result in some
+  //        difficult-to-analyze code and make pruning branches much harder.
+  //        This is a problem if the switch expression itself can be restricted
+  //        by inlining or CVP.
```
Thanks for looking into the benchmark and reviewing. I decided to guard the bitmap creation with ConvertSwitchToLookup for now instead of completely removing the ConvertSwitchToLookup change. Hopefully the new benchmark results will show no other regressions, so the other kinds of transformations can still be applied earlier without issues, otherwise I will remove it for now.

OutOfCache committed:
SimplifyCFG: Collect lookup table statistics at creation

Since the SwitchReplacement is created before we decide to replace the
switch, only increase the table counters when we actually create them.

OutOfCache committed:
Revert "SimplifyCFG: Delay creation of bitmaps"

This reverts commit 27179dc87b8b574f87697c7a4ac622e374559348.

OutOfCache committed:
Revert "SimplifyCFG: Delay check for switch-to-lookup option"

This reverts commit 081d8503d0430f6a3a150b1c0a8d7de27f2b4b63.

nikic reviewed (approved):
LGTM
